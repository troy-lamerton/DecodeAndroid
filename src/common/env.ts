import { config } from 'dotenv';
import * as path from 'path';

config({ path: path.resolve(process.cwd(), '.env.local') });
config();

export interface IGlobals {
    TESTING: boolean;
    SYMBOLS_S3_BUCKET: string;
    SYMBOLS_S3_ACCESS_KEY: string;
    SYMBOLS_S3_SECRET_ACCESS_KEY: string;
}

export const Globals: IGlobals = {
    TESTING: Boolean(global['TESTING']), // only set by jest config, not env args
    SYMBOLS_S3_BUCKET: String(global['SYMBOLS_S3_BUCKET'] || process.env['SYMBOLS_S3_BUCKET']),
    SYMBOLS_S3_ACCESS_KEY: String(global['SYMBOLS_S3_ACCESS_KEY'] || process.env['SYMBOLS_S3_ACCESS_KEY']),
    SYMBOLS_S3_SECRET_ACCESS_KEY: String(global['SYMBOLS_S3_SECRET_ACCESS_KEY'] || process.env['SYMBOLS_S3_SECRET_ACCESS_KEY']),
};
