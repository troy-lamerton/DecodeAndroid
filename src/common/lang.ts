import { isUndefined, get, negate, isEmpty } from 'lodash';

export function callElseThrow<T>(func: undefined | (() => T), throws: any) {
    if (isUndefined(func)) {
        throw throws;
    }
    return func();
}

export const TODO = (msg?: string): never => {
    throw `[TODO] ${msg || 'not yet implemented'}`;
};

export function randomHex(n: number): string {
    let rs = '';
    let r = n % 8,
        q = (n - r) / 8,
        i;
    for (i = 0; i < q; i++) {
        rs += Math.random()
            .toString(16)
            .slice(2);
    }
    if (r > 0) {
        rs += Math.random()
            .toString(16)
            .slice(2, i);
    }
    return rs;
}

export const notEmpty = negate(isEmpty);

export const isDefined = <T>(value: any): value is T => value !== undefined;
