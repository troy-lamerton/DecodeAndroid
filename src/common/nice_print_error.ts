export function prettyPrintError(err: any): string {
    const output = serializeError(err, 1000);
    console.error('[ERROR]', output);
    return output;
}

export function serializeError(value: any, maxLength = 2048): string {
    if (typeof value === 'string') {
        return value.slice(0, maxLength);
    }

    if (typeof value === 'object') {
        return JSON.stringify(destroyCircular(value, [])).slice(0, maxLength);
    }

    // People sometimes throw things besides Error objectsâ€¦
    if (typeof value === 'function') {
        // JSON.stringify discards functions. We do too, unless a function is thrown directly.
        return `[Function: ${value.name || 'anonymous'}]`.slice(0, maxLength);
    }

    return value;
}

// @See(https://github.com/sindresorhus/serialize-error/blob/master/index.js)
const destroyCircular = (from: any, seen: any): object => {
    const to = Array.isArray(from) ? [] : {};

    seen.push(from);

    // TODO: Use `Object.entries() when targeting Node.js 8
    for (const key of Object.keys(from)) {
        const value = from[key];

        if (typeof value === 'function') {
            continue;
        }

        if (!value || typeof value !== 'object') {
            to[key] = value;
            continue;
        }

        if (!seen.includes(from[key])) {
            to[key] = destroyCircular(from[key], seen.slice());
            continue;
        }

        to[key] = '[Circular]';
    }

    const commonProperties = ['name', 'message', 'stack', 'code'];

    for (const property of commonProperties) {
        if (typeof from[property] === 'string') {
            to[property] = from[property];
        }
    }

    return to;
};
