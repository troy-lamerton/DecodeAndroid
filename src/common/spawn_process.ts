import { onExit, readableToString, streamEnd, streamWrite } from '@rauschma/stringio';
import { spawn, spawnSync } from 'child_process';
import { Writable } from 'stream';
import l from './logger';

/**
 * test me
 * Line A: We ignore stdin, want to access stdout via a stream and forward stderr to process.stderr.
 * Line B: We await until echoReadable() is completely done. Without this await, DONE would be printed before the first line of source.stdout.
 */
export async function demo_readProcessOutput(inputFile: string): Promise<string> {
    const source = spawn('cat', [inputFile], { stdio: ['ignore', 'pipe', process.stderr] }); // (A)

    return await readableToString(source.stdout!);
}

export async function demo_writeProcessInput(linesToWrite = ['First line', 'Second line']) {
    const sink = spawn('cat', [], { stdio: ['pipe', process.stdout, process.stderr] }); // (A)

    await writeToWritable(sink.stdin!, linesToWrite); // (B)
    await onExit(sink);
}

export function runProcessForOutput(executableFile: string, args: string[]): [string, string] {
    l.info(`running: ${executableFile} ${args.join(' ')}`);
    const proc = spawnSync(executableFile, args, { stdio: ['ignore', 'pipe', 'pipe'] });

    const code = proc.status;

    const errorOutput = proc.stderr.toString().trim();

    if (code === 0) {
        if (errorOutput) l.w(errorOutput);
        return [proc.stdout.toString().trimRight(), errorOutput];
    } else {
        throw errorOutput;
    }
}

export async function writeToProcessForOutput(
    executableFile: string,
    args: string[],
    inputLines: string[],
): Promise<string> {
    l.trace(`${executableFile} ${args.join(' ')}`);

    const proc = spawn(executableFile, args, { stdio: ['pipe', 'pipe', process.stderr] });

    l.trace('writing ' + inputLines.join('\n'));

    await writeToWritable(proc.stdin!, inputLines, true);
    l.trace('writing to child process done');
    const output = await readableToString(proc.stdout!);
    l.trace('read to end done');

    // if (!proc.killed) await onExit(proc);
    proc.kill();
    l.trace('process did exit done');

    return output.trimRight();
}

async function writeToWritable(writable: Writable, lines: string[], instantEnd = true) {
    for (const line of lines) {
        await streamWrite(writable, line + '\n');
    }
    if (instantEnd) await streamEnd(writable);
}
