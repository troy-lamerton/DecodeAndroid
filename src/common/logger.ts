import Pino, { Logger } from 'pino';
import { serializeError } from './nice_print_error';
import { Globals } from './env';

const tag = `crashdecoder_android`;

const level = process.env.LOG_LEVEL || 'debug';
// logging that conforms to cos-* elastic search index
const humanFormat: Pino.LoggerOptions = {
    base: {},
    prettyPrint: {
        translateTime: 'SYS:HH:MM:ss.l',
        colorize: true,
    },
    level,
};
const elasticFormat: Pino.LoggerOptions = {
    level,
    timestamp: () => `,"@timestamp":"${new Date().toJSON()}"`,
    changeLevelName: 'log_level',
    base: {
        tag,
    },
};

const printForHuman =
    process.env.HUMAN_LOGS || process.env.NODE_ENV === 'development' || Globals.TESTING;
const loggerConfig: Pino.LoggerOptions = printForHuman ? humanFormat : elasticFormat;

type Best = {
    w(msg: string, stat?: string): void;
    e(msg: any, stat?: string): void;
};

const l = Pino(loggerConfig) as (Logger & Best);

const defaults = (stat: string, logtype: string, merger: any = {}) => {
    return printForHuman
        ? {
              ...merger,
          }
        : {
              stat,
              logtype,
              ...merger,
          };
};

l.w = (msg: string, stat?: string) => {
    l.warn(defaults(stat || 'log', 'Warning', { msg }));
};

l.e = (err: any, stat?: string) => {
    l.error(
        defaults(stat || 'log', 'Error', {
            is_error: true,
            msg: serializeError(err),
        }),
    );
};

export default l;
