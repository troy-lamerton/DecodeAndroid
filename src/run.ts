import './common/env';
import { CrashReportV2Body } from './schema/crashes_v2/crashes_v2.schema';
import { CrashReportV2 } from './processing_v2/report_v2';
import { validateAndProcess } from './processing_v2/start_processing';
import { setS3ClientConfig } from './processing_v2/s3_client';
import { serializeError } from './common/nice_print_error';
import l from './common/logger';
import md5 from 'md5';
import { S3 } from 'aws-sdk';

type AnyObject = { [key: string]: any };

export async function decodeJsonPayload(jsonPayload: string, s3Config?: S3.Types.ClientConfiguration): Promise<AnyObject | null> {
    if (s3Config) setS3ClientConfig(s3Config)

    const payloadBody = JSON.parse(jsonPayload) as CrashReportV2Body;
    const hash = md5(payloadBody.report_contents);
    const report = new CrashReportV2('init', payloadBody, hash);
    try {
        return await validateAndProcess(report);
    } catch (fatalError) {
        report.markDone(serializeError(fatalError), true);
        l.e(report.descriptionSecret);
        // null to indicate a fatal error
        return null;
    }
}
