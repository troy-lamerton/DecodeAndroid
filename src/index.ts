import { decodeJsonPayload } from './run';

(async function() {
    console.log('hello world');
    console.log('decode function: ' + decodeJsonPayload.toString())
})();
