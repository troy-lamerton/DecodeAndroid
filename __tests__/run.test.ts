import { decodeJsonPayload } from '../src/run';
import path from 'path';
import fs from 'fs-extra';
import { S3 } from 'aws-sdk';
import { Globals } from '../src/common/env';

describe('run', () => {
    const dummy = {
        // native crash from an arm64 build
        idlegame_report_arm64_native: {
            files: {
                request_body_json: path.join(
                    __dirname,
                    'dummy_data',
                    'requestBody_2_native.json',
                ),
                // when you have access, these symbols can be downloaded from:
                // https://s3.console.aws.amazon.com/s3/buckets/gitlab-test-symbols/symbols/android/f39fefb7-01d9-4509-8f76-4a99957dcf24/ARM64/?region=us-east-2
                symbols_libil2cpp: path.join(
                    __dirname,
                    'dummy_data',
                    'test_symbols',
                    'libil2cpp.so.debug',
                ),
            },
        },
        idlegame_report_arm64_jvm: {
            files: {
                request_body_json: path.join(__dirname, 'dummy_data', 'requestBody_1_jvm.json'),
            },
        },
    };

    async function getNativePayloadContents(): Promise<string> {
        const jsonFileContents = await fs.readFile(
            dummy.idlegame_report_arm64_native.files.request_body_json,
        );
        return jsonFileContents.toString();
    }

    async function getJvmPayloadContents(): Promise<string> {
        const jsonFileContents = await fs.readFile(
            dummy.idlegame_report_arm64_jvm.files.request_body_json,
        );
        return jsonFileContents.toString();
    }

    test('ensure s3 config is set', () => {
        const expectedEnvVars = [
            'SYMBOLS_S3_ACCESS_KEY',
            'SYMBOLS_S3_SECRET_ACCESS_KEY'
        ]
        const msg = `To run the tests that use s3, please create a file called '.env.local' next to the package.json.
                     It should contain:\n${expectedEnvVars.map(envName => `${envName}=?`).join('\n')}`.replace(/ +/g, ' ')
        expectedEnvVars.forEach((envName) => {
            const envValue = process.env[envName];
            if (!envValue) {
                throw msg;
            }
        })
    })

    test('native payload', async () => {
        const nativeJsonPayload = await getNativePayloadContents();
        const s3Config: S3.Types.ClientConfiguration = {
            credentials: {
                accessKeyId: Globals.SYMBOLS_S3_ACCESS_KEY,
                secretAccessKey: Globals.SYMBOLS_S3_SECRET_ACCESS_KEY,
            },
            region: process.env.AWS_DEFAULT_REGION,
        }
        const output = await decodeJsonPayload(nativeJsonPayload, s3Config);
        console.log(output);
        if (!output) {
            return fail('output is falsy');
        }
        expect(output.decoder_error_count!).toBe(0);
        delete output['@timestamp']
        delete output['decoder_processing_duration']
        expect(output).toMatchSnapshot();
    }, 30 * 10 * 1000);

    test('jvm payload', async () => {
        const jvmJsonPayload = await getJvmPayloadContents();
        const s3Config: S3.Types.ClientConfiguration = {
            credentials: {
                accessKeyId: Globals.SYMBOLS_S3_ACCESS_KEY,
                secretAccessKey: Globals.SYMBOLS_S3_SECRET_ACCESS_KEY,
            },
            region: process.env.AWS_DEFAULT_REGION,
        }
        const output = await decodeJsonPayload(jvmJsonPayload, s3Config);
        console.log(output);
        if (!output) {
            return fail('output is falsy');
        }
        expect(output.decoder_error_count!).toBe(0);
        // ensure snapshot doesnt change based on when the test is run
        delete output['@timestamp']
        delete output['decoder_processing_duration']
        expect(output).toMatchSnapshot();
    }, 10 * 1000);
});
