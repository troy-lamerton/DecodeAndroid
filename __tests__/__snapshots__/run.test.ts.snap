// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run jvm payload 1`] = `
Object {
  "client_buttons_pressed": Array [],
  "client_commit": "6cd428aa",
  "client_logs": Array [],
  "client_version": "7.0.0+6cd428aa",
  "crash_contents_md5": "614adfa483f4ef898ff5e96496ce6c0c",
  "crashreport_description": "#00 [vdso]. in _unknown_file:0
#01 libc.tgkill in tgkill:4
#02 libunity. in _unknown_file:0
#03 libil2cpp. in _unknown_file:0
#04 libil2cpp. in _unknown_file:0
#05 libil2cpp. in _unknown_file:0
#06 libil2cpp. in _unknown_file:0
#07 libil2cpp. in _unknown_file:0
#08 libil2cpp. in _unknown_file:0
#09 libil2cpp. in _unknown_file:0
#10 libil2cpp. in _unknown_file:0
#11 libil2cpp. in _unknown_file:0
#12 libil2cpp. in _unknown_file:0
#13 libil2cpp. in _unknown_file:0
#14 libil2cpp. in _unknown_file:0
#15 libil2cpp. in _unknown_file:0
#16 libil2cpp. in _unknown_file:0
#17 libil2cpp. in _unknown_file:0
#18 libil2cpp. in _unknown_file:0
#19 libil2cpp. in _unknown_file:0
#20 libil2cpp. in _unknown_file:0
#21 libil2cpp. in _unknown_file:0
#22 libunity. in _unknown_file:0
#23 libunity. in _unknown_file:0
#24 libunity. in _unknown_file:0
#25 libunity. in _unknown_file:0
#26 libunity. in _unknown_file:0
#27 libunity. in _unknown_file:0
#28 libunity. in _unknown_file:0
#29 libunity. in _unknown_file:0
#30 libunity. in _unknown_file:0
#31 base. in _unknown_file:0",
  "crashreport_id": "0458c555e67a440487759ff2e00aa36e",
  "crashreport_reason": "",
  "created_at": "2020-05-18T15:20:50.126542Z",
  "decoder_debug": "Build: test - Android
Version: 7.0.0+6cd428aa (commit: 6cd428aa)
Person: client_KQxztP79rS5cmjeKkTxIfKpU. userId: KQxztP79rS5cmjeKkTxIfKpU",
  "decoder_error_count": 0,
  "decoder_errors": undefined,
  "decoder_name": "V2.4.0 - new crashreporting",
  "decoder_status": "success",
  "decoder_version": 20200516,
  "device_aspectRatio": "18x9",
  "device_graphicsAPI": "OpenGLES3",
  "device_graphicsMemory": 1024,
  "device_graphicsProcessor": "Adreno (TM) 508",
  "device_graphicsVendor ": "Qualcomm",
  "device_memory": 3747,
  "device_model": "motorola moto g(6) plus",
  "device_os": "Android OS 9 / API-28 (PPWS29.116-20-6/e70a45)",
  "device_processor": "ARM64 FP ASIMD AES",
  "device_resolution": "2160x1080",
  "hostId": "client_4c1e9e25ac660d532cdbe3a494ec7698",
  "id": "12365414408322c0XI8w",
  "person_id": "client_KQxztP79rS5cmjeKkTxIfKpU. userId: KQxztP79rS5cmjeKkTxIfKpU",
  "release_stage": "test",
  "stat": "client.crashreport.native.android",
  "timestamp": "2020-05-18T15:20:50.1289Z",
  "user": "client_KQxztP79rS5cmjeKkTxIfKpU",
  "version": "7.0.0+6cd428aa",
}
`;

exports[`run native payload 1`] = `
Object {
  "client_buttons_pressed": Array [],
  "client_commit": "6cd428aa",
  "client_logs": Array [],
  "client_version": "7.0.0+6cd428aa",
  "crash_contents_md5": "f6ffe8c7d1702f7f4b4d2e89a1ea235d",
  "crashreport_description": "pid: 12300, tid: 12300, name: Thread  >>> app.app <<<
#00  6f128 /system/lib64/libc.so
#01 3c8750 <app-lib>/arm64/libunity.so: Routine void UnityGUID::Transfer<StreamedBinaryWrite>(StreamedBinaryWrite&) at ??:?
#02 2897e54 <app-lib>/arm64/libil2cpp.so: Routine RuntimeInvoker_TrueVoid_t22962CB4...017_Int32_t585191389E07734F19F3156FF88FB3EF4800D102(void (*)(), MethodInfo const*, void*, void**) at il2cppOutput/Il2CppInvokerTable.cpp:141169
#03 2b474e0 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/Runtime.cpp:597
#04 2b4c4e8 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::vm::Runtime::InvokeConvertArgs(MethodInfo const*, void*, Il2CppObject**, int, Il2CppException**) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/Runtime.cpp:?
#05 2b273b0 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::icalls::mscorlib::System::Reflection::MonoMethod::InternalInvoke(Il2CppReflectionMethod*, Il2CppObject*, Il2CppArray*, Il2CppException**) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\icalls\\\\mscorlib\\\\System.Reflection/MonoMethod.cpp:242
#06 5848c0c <app-lib>/arm64/libil2cpp.so: Routine MonoMethod_InternalInvoke_mB9735487...789 at il2cppOutput/mscorlib11.cpp:29139
#07 2fc41c0 <app-lib>/arm64/libil2cpp.so: Routine il2cpp_codegen_get_virtual_invoke_data(unsigned int, Il2CppObject const*) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\codegen/il2cpp-codegen-il2cpp.h:305
#08 5a6efe8 <app-lib>/arm64/libil2cpp.so: Routine UnityAction_Invoke_mC9FF5AA1...71A at il2cppOutput/UnityEngine.CoreModule.cpp:28937
#09 5a4ebcc <app-lib>/arm64/libil2cpp.so: Routine UnityAction_1_Invoke_m3685D48F...92B_gshared at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/ClassInlines.h:?
#10 37ae3ec <app-lib>/arm64/libil2cpp.so: Routine InputService_ExecuteHandlerSafe_mF9232018...150 at il2cppOutput/Main82.cpp:61234
#11 37ae200 <app-lib>/arm64/libil2cpp.so: Routine InputService_HandleEvent_m31C7E8D1...5F6 at il2cppOutput/Main82.cpp:60725
#12 38cf348 <app-lib>/arm64/libil2cpp.so: Routine U3CU3Ec__DisplayClass0_0_U3C_ctorU3Eb__1_m76DF88E2...D14 at il2cppOutput/Main68.cpp:52575
#13 4b74634 <app-lib>/arm64/libil2cpp.so: Routine Action_1_Invoke_mB86FC1B3...1D5_gshared at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/ClassInlines.h:?
#14 514e354 <app-lib>/arm64/libil2cpp.so: Routine InlineExecuteAllSystem_Execute_m2A89F780...4CA at il2cppOutput/Helper.cpp:29406
#15 39f64d4 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::utils::LeaveTargetStack::push(int) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\utils/LeaveTargetStack.h:22
#16 3699b80 <app-lib>/arm64/libil2cpp.so: Routine il2cpp_codegen_get_virtual_invoke_data(unsigned int, Il2CppObject const*) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\codegen/il2cpp-codegen-il2cpp.h:305
#17 3c617d8 <app-lib>/arm64/libil2cpp.so: Routine ClientController_ControllerUpdate_m86390D47...BED at il2cppOutput/Main26.cpp:56324
#18 36998d8 <app-lib>/arm64/libil2cpp.so: Routine ControllerBridge_t316E85A0...F02::get_controller_4() const at il2cppOutput/Main37.cpp:34416
#19 2897c38 <app-lib>/arm64/libil2cpp.so: Routine RuntimeInvoker_TrueVoid_t22962CB4...017(void (*)(), MethodInfo const*, void*, void**) at il2cppOutput/Il2CppInvokerTable.cpp:141106
#20 2b474e0 <app-lib>/arm64/libil2cpp.so: Routine il2cpp::vm::Runtime::Invoke(MethodInfo const*, void*, void**, Il2CppException**) at C:\\\\Program Files\\\\Unity\\\\Hub\\\\Editor\\\\2019.2.8f1\\\\Editor\\\\Data\\\\il2cpp\\\\libil2cpp\\\\vm/Runtime.cpp:597
#21 45187c <app-lib>/arm64/libunity.so: Routine void std::__adjust_heap<Experimental::GlobalIllumination::LightDataGI*, long, Experimental::GlobalIllumination::LightDataGI, __gnu_cxx::__ops::_Iter_comp_iter<SortByHashPred<Experimental::GlobalIllumination::LightDataGI, SceneLights::Hasher> > >(Experimental::GlobalIllumination::LightDataGI*, long, long, Experimental::GlobalIllumination::LightDataGI, __gnu_cxx::__ops::_Iter_comp_iter<SortByHashPred<Experimental::GlobalIllumination::LightDataGI, SceneLights::Hasher> >) at ??:?
#22 441cd8 <app-lib>/arm64/libunity.so: Routine GfxDeviceWorker::RunCommand(ThreadedStreamBuffer&) at ??:?
#23 43c6d4 <app-lib>/arm64/libunity.so: Routine std::_Rb_tree<GfxRasterState, std::pair<GfxRasterState const, ClientDeviceRasterState>, std::_Select1st<std::pair<GfxRasterState const, ClientDeviceRasterState> >, MemoryCompLess<GfxRasterState>, std::allocator<std::pair<GfxRasterState const, ClientDeviceRasterState> > >::_M_get_insert_unique_pos(GfxRasterState const&) at ??:?
#24 56f010 <app-lib>/arm64/libunity.so: Routine CachingManager::GetDefaultCache() at ??:?
#25 4cfd4c <app-lib>/arm64/libunity.so: Routine void std::__introsort_loop<int*, long, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter> >(int*, int*, long, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter>) at ??:?
#26 4cfd80 <app-lib>/arm64/libunity.so: Routine void std::__final_insertion_sort<int*, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter> >(int*, int*, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter>) at ??:?
#27 4cff54 <app-lib>/arm64/libunity.so: Routine void std::__adjust_heap<int*, long, int, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter> >(int*, long, long, int, __gnu_cxx::__ops::_Iter_comp_iter<IndexSorter>) at ??:?
#28 3e0ad4 <app-lib>/arm64/libunity.so: Routine void audio::mixer::GroupConstant::Transfer<SafeBinaryRead>(SafeBinaryRead&) at ??:?
#29 3da8f4 <app-lib>/arm64/libunity.so: Routine AudioSource::UpdateLocalChannels(dynamic_array<SoundChannel, 0ul>&) at ??:?
#30  648f0 /data/app/com.gamingforgood.clashofstreamers-5mNSCmwAJ51ek4Mdbjy-zw==/oat/arm64/base.odex",
  "crashreport_id": "4e2be758209d4ab8954f7c06ff2481c5",
  "crashreport_reason": "",
  "created_at": "2020-05-18T15:35:50.756736Z",
  "decoder_debug": "Build: test - Android
Version: 7.0.0+6cd428aa (commit: 6cd428aa)
Person: client_KQxztP79rS5cmjeKkTxIfKpU. userId: KQxztP79rS5cmjeKkTxIfKpU",
  "decoder_error_count": 0,
  "decoder_errors": undefined,
  "decoder_name": "V2.4.0 - new crashreporting",
  "decoder_status": "success",
  "decoder_version": 20200516,
  "device_aspectRatio": "18x9",
  "device_graphicsAPI": "OpenGLES3",
  "device_graphicsMemory": 1024,
  "device_graphicsProcessor": "Adreno (TM) 508",
  "device_graphicsVendor ": "Qualcomm",
  "device_memory": 3747,
  "device_model": "motorola moto g(6) plus",
  "device_os": "Android OS 9 / API-28 (PPWS29.116-20-6/e70a45)",
  "device_processor": "ARM64 FP ASIMD AES",
  "device_resolution": "2160x1080",
  "hostId": "client_4c1e9e25ac660d532cdbe3a494ec7698",
  "id": "12374420708422V1Xnc6",
  "person_id": "client_KQxztP79rS5cmjeKkTxIfKpU. userId: KQxztP79rS5cmjeKkTxIfKpU",
  "release_stage": "test",
  "stat": "client.crashreport.native.android",
  "timestamp": "2020-05-18T15:35:50.75891Z",
  "user": "client_KQxztP79rS5cmjeKkTxIfKpU",
  "version": "7.0.0+6cd428aa",
}
`;
